mailbox/postgres/src/main/java/org/apache/james/mailbox/postgres/mail/dao/PostgresMailboxMessageDAO.java

```diff
@@ -0,0 +1,420 @@
+
+    public PostgresMailboxMessageDAO(PostgresExecutor postgresExecutor) {
+        this.postgresExecutor = postgresExecutor;
+        this.threadIdFactory = new ThreadId.Factory(new PostgresMessageId.Factory());
+    }
+
+    public Mono<MessageUid> findFirstUnseenMessageUid(PostgresMailboxId mailboxId) {
+        return postgresExecutor.executeRow(dslContext -> Mono.from(selectMessageUidByMailboxIdAndExtraConditionQuery(mailboxId,
+                IS_SEEN.eq(false), Optional.of(1), dslContext)))
+            .map(RECORD_TO_MESSAGE_UID_FUNCTION);
+    }
+
+    public Flux<MessageUid> findAllRecentMessageUid(PostgresMailboxId mailboxId) {
+        return postgresExecutor.executeRows(dslContext -> Flux.from(selectMessageUidByMailboxIdAndExtraConditionQuery(mailboxId,
+                IS_SEEN.eq(true), Optional.empty(), dslContext)))
+            .map(RECORD_TO_MESSAGE_UID_FUNCTION);
+    }
+
+    public Flux<MessageUid> listAllMessageUid(PostgresMailboxId mailboxId) {
+        return postgresExecutor.executeRows(dslContext -> Flux.from(selectMessageUidByMailboxIdAndExtraConditionQuery(mailboxId,
+                DSL.noCondition(), Optional.empty(), dslContext)))
+            .map(RECORD_TO_MESSAGE_UID_FUNCTION);
+    }
+
+    public Mono<MessageUid> findLastMessageUid(PostgresMailboxId mailboxId) {
+        return postgresExecutor.executeRow(dslContext -> Mono.from(dslContext.select(MESSAGE_UID)
+                .from(TABLE_NAME)
+                .where(MAILBOX_ID.eq(mailboxId.asUuid()))
+                .orderBy(MESSAGE_UID.desc())
+                .limit(1)))
+            .map(RECORD_TO_MESSAGE_UID_FUNCTION);
+    }
+
+    public Mono<Void> deleteByMailboxId(PostgresMailboxId mailboxId) {
+        return postgresExecutor.executeVoid(dslContext -> Mono.from(dslContext.deleteFrom(TABLE_NAME)
+            .where(MAILBOX_ID.eq(mailboxId.asUuid()))));
+    }
+
+    public Mono<Void> deleteByMailboxIdAndMessageUid(PostgresMailboxId mailboxId, MessageUid messageUid) {
+        return postgresExecutor.executeVoid(dslContext -> Mono.from(dslContext.deleteFrom(TABLE_NAME)
+            .where(MAILBOX_ID.eq(mailboxId.asUuid()))
+            .and(MESSAGE_UID.eq(messageUid.asLong()))));
+    }
+
+    public Mono<Void> deleteByMailboxIdAndMessageUids(PostgresMailboxId mailboxId, List<MessageUid> uids) {
+        return postgresExecutor.executeVoid(dslContext -> Mono.from(dslContext.deleteFrom(TABLE_NAME)
+            .where(MAILBOX_ID.eq(mailboxId.asUuid()))
+            .and(MESSAGE_UID.in(uids.stream().map(MessageUid::asLong).toArray(Long[]::new)))));
+    }
+
+    public Mono<Void> deleteByMailboxIdAndMessageUidBetween(PostgresMailboxId mailboxId, MessageUid from, MessageUid to) {
+        return postgresExecutor.executeVoid(dslContext -> Mono.from(dslContext.deleteFrom(TABLE_NAME)
+            .where(MAILBOX_ID.eq(mailboxId.asUuid()))
+            .and(MESSAGE_UID.greaterOrEqual(from.asLong()))
+            .and(MESSAGE_UID.lessOrEqual(to.asLong()))));
```

Comment:
Systematically use `RETURNING`. We need to be able to build `MessageMetaData` information solely from the information returned from `message_mailbox`. (To that end we need to add the "size" parameter onto the `message_mailbox` table)